src/modules/ekf2/EKF/output_predictor.cpp

```diff
@@ -242,137 +293,258 @@ void OutputPredictor::calculateOutputStates(const uint64_t time_us, const Vector
+			continue;
+		}
 
-	// convert the quaternion delta to a delta angle
-	const float scalar = (q_error(0) >= 0.0f) ? -2.f : 2.f;
+		if (!_output_filter_aligned) {
 
-	const Vector3f delta_ang_error{scalar * q_error(1), scalar * q_error(2), scalar * q_error(3)};
+			// calculate the quaternion rotation delta from the EKF to output observer states at the EKF fusion time horizon
+			const Quatf q_delta((quat_state * output_delayed.quat_nominal.inversed()).normalized());
+			const Dcmf rot_change{q_delta};
+
+			// add the reset amount to the output observer buffered data
+			for (uint8_t i = 0; i < _output_buffer.get_length(); i++) {
+
+				_output_buffer[i].quat_nominal = (q_delta * _output_buffer[i].quat_nominal).normalized();
+
+				// apply change to velocity
+				_output_buffer[i].vel = rot_change * _output_buffer[i].vel;
+				_output_buffer[i].vel_alternative = rot_change * _output_buffer[i].vel_alternative;
+			}
+
+			// apply the change in attitude quaternion to our newest quaternion estimate
+			_output_new.quat_nominal = (q_delta * _output_new.quat_nominal).normalized();
 
-	// calculate a gain that provides tight tracking of the estimator attitude states and
-	// adjust for changes in time delay to maintain consistent damping ratio of ~0.7
-	const uint64_t time_latest_us = _time_last_update_states_us;
-	const float time_delay = fmaxf((time_latest_us - time_delayed_us) * 1e-6f, _dt_update_states_avg);
-	const float att_gain = 0.5f * _dt_update_states_avg / time_delay;
 
-	// calculate a corrrection to the delta angle
-	// that will cause the INS to track the EKF quaternions
-	_delta_angle_corr = delta_ang_error * att_gain;
-	_output_tracking_error(0) = delta_ang_error.norm();
+			// apply change to velocity
+			_output_new.vel = rot_change * _output_new.vel;
+			_output_new.vel_alternative = rot_change * _output_new.vel_alternative;
 
-	/*
-	* Loop through the output filter state history and apply the corrections to the velocity and position states.
-	* This method is too expensive to use for the attitude states due to the quaternion operations required
-	* but because it eliminates the time delay in the 'correction loop' it allows higher tracking gains
-	* to be used and reduces tracking error relative to EKF states.
-	*/
+			propagateVelocityUpdateToPosition();
 
-	// Complementary filter gains
-	const float vel_gain = _dt_correct_states_avg / math::constrain(_vel_tau, _dt_correct_states_avg, 10.f);
-	const float pos_gain = _dt_correct_states_avg / math::constrain(_pos_tau, _dt_correct_states_avg, 10.f);
 
-	// calculate down velocity and position tracking errors
-	const float vert_vel_err = (vel_state(2) - output_vert_delayed.vert_vel);
-	const float vert_vel_integ_err = (pos_state(2) - output_vert_delayed.vert_vel_integ);
+			// calculate the velocity and position deltas between the output and EKF at the EKF fusion time horizon
+			output_delayed.vel = q_delta.rotateVectorInverse(output_delayed.vel);
```

Comment:
Shouldn't we use `rotateVector(...` or `rot_change * ...` here?
Also, it seems that this is only needed because the velocity is corrected during the quaternion correction above.